README
libDXF is a library with DXF related functions.

Currently this project has a “Work in progress” status.

The code base is still incomplete and is not debugged to an acceptable level.

You are always welcome to help with patches or bug reports :-)


ChangeLog
commit 4904c23c48a9175cb1f5de8092625b96ac06d7eb
Author: bert <bert.timmerman@xs4all.nl>
Date:   Fri Jul 17 07:50:39 2015 +0200

    HATCH: do some basic checks in dxf_hatch_pattern_def_line_init().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit d65a4876ace70e28d8c9772e623a15a849053567
Author: bert <bert.timmerman@xs4all.nl>
Date:   Fri Jul 17 07:45:48 2015 +0200

    HATCH: do some basic checks in dxf_hatch_pattern_init().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit d3e489ddf66772acabd1e7d690221612c7fba0c6
Author: bert <bert.timmerman@xs4all.nl>
Date:   Thu Jul 16 19:29:15 2015 +0200

    HEADER: added dxf_header_free().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 4badce461c6df27616a3191b6384424a889fb5e8
Author: bert <bert.timmerman@xs4all.nl>
Date:   Thu Jul 16 06:02:30 2015 +0200

    HEADER: clean up before return in dxf_header_write().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit f0b82d2b585de46565f83f620902fd5fe9ab1ef6
Author: bert <bert.timmerman@xs4all.nl>
Date:   Thu Jul 16 00:31:52 2015 +0200

    HEADER: added basic tests to dxf_header_read() and changed the return type.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 6cf537f9c1915120a8519b364316660ad622c07b
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 23:44:20 2015 +0200

    HEADER: added basic tests to dxf_header_read_parser().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 7b0574ffe9aaed54143aea3238d9efad3af5abd3
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 23:38:09 2015 +0200

    HEADER: added basic tests to dxf_header_read_parse_n_double().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit f1d2698b548fa7290cf01672a8bb5a8fc2161795
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 23:35:48 2015 +0200

    HEADER: added basic tests to dxf_header_read_parse_int().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 7c828e36f00343602ad15516786234478f004b76
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 22:39:53 2015 +0200

    HEADER: added basic tests to dxf_header_write().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 0369207f50721dba77f9d899d96095fa3b7cb92b
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 22:35:32 2015 +0200

    HEADER: test for a NULL pointer in dxf_header_write_metric_new().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 5fdca44ec7ff8d662e72b939ddbf55c9ed75fa77
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 22:15:20 2015 +0200

    HEADER: added basic tests to dxf_header_init() and changed the return type.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit a824e77c028208eefe133023d468e101e362cce2
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 21:18:17 2015 +0200

    HEADER: test for a NULL pointer in dxf_header_acad_version_from_string().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 36aa8381e8ebb0885cd9cbece3e02bdad4d6f0ec
Author: bert <bert.timmerman@xs4all.nl>
Date:   Wed Jul 15 08:20:48 2015 +0200

    COLOR: added dxf_RGB_color_free_chain().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 10c86d075af85dc43d480b5d1bfee68103a47e69
Author: bert <bert.timmerman@xs4all.nl>
Date:   Tue Jul 14 23:01:36 2015 +0200

    COLOR: added a "next" member to the DxfRGBColor struct.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 98653239e516e7b900ecbff9ddb65e88ef67d1d2
Author: bert <bert.timmerman@xs4all.nl>
Date:   Tue Jul 14 22:53:36 2015 +0200

    COMMENT: correction in the Doxygen header.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit c591686cc5bc61c5868fe5787bb48e1cffa5f238
Author: bert <bert.timmerman@xs4all.nl>
Date:   Tue Jul 14 22:51:10 2015 +0200

    COMMENT: added dxf_comment_get_value().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 8246d32cc2d02416877b33c907e76dec5b9a67b9
Author: bert <bert.timmerman@xs4all.nl>
Date:   Tue Jul 14 00:04:36 2015 +0200

    COMMENT: added dxf_comment_set_value().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit cea790429923b7fc92e63647694b5995c81c0c4a
Author: bert <bert.timmerman@xs4all.nl>
Date:   Mon Jul 13 20:36:11 2015 +0200

    COMMENT: corrections in Doxygen comments.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 74dd2b1486bca87e83dc192b36be1f36118c4f58
Author: bert <bert.timmerman@xs4all.nl>
Date:   Mon Jul 13 08:01:12 2015 +0200

    COMMENT: added dxf_comment_free_chain().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 74927c6c606922ab5fd34d156f3fe4de76923f06
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sun Jul 12 23:36:33 2015 +0200

    XRECORD: corrections in Doxygen comments.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 208ec2e56c6808dc119b434ef8d51426e8a6e78c
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sun Jul 12 22:51:48 2015 +0200

    XRECORD: added dxf_xrecord_free_chain().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit b304aa212db6260e2f2d248213518b11a473d70c
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sun Jul 12 22:08:51 2015 +0200

    XLINE: corrections in Doxygen comments.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit dbc1d27da131d958896c30a78bea217553608d17
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sun Jul 12 18:01:35 2015 +0200

    XLINE: added dxf_xline_free_chain().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 60055483b4026bb05adb90d4171e39557488c6e9
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sun Jul 12 08:55:28 2015 +0200

    VPORT: corrections in Doxygen comments.
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>

commit 26539949b24da074a6332067c3715b59547d612f
Author: bert <bert.timmerman@xs4all.nl>
Date:   Sat Jul 11 23:43:50 2015 +0200

    VPORT: added dxf_vport_free_chain().
    
    Signed-off-by: bert <bert.timmerman@xs4all.nl>


src/dim.h
/*!
 * \file dim.h
 *
 * \author Copyright (C) 2008, 2012 by Bert Timmerman <bert.timmerman@xs4all.nl>.
 *
 * \brief Definition of a DXF dimension entity (\c DIM).
 *
 * <hr>
 * <h1><b>Copyright Notices.</b></h1>\n
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.\n\n
 * This program is distributed in the hope that it will be useful, but
 * <b>WITHOUT ANY WARRANTY</b>; without even the implied warranty of
 * <b>MERCHANTABILITY</b> or <b>FITNESS FOR A PARTICULAR PURPOSE</b>.\n
 * See the GNU General Public License for more details.\n\n
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to:\n
 * Free Software Foundation, Inc.,\n
 * 59 Temple Place,\n
 * Suite 330,\n
 * Boston,\n
 * MA 02111 USA.\n
 * \n
 * Drawing eXchange Format (DXF) is a defacto industry standard for the
 * exchange of drawing files between various Computer Aided Drafting
 * programs.\n
 * DXF is an industry standard designed by Autodesk(TM).\n
 * For more details see http://www.autodesk.com.
 * <hr>
 */


#ifndef _DIM_H
#define _DIM_H


#include "global.h"


/*!
 * \brief DXF definition of an AutoCAD dimension.
 */
typedef struct
dxf_dimension
{
        /* Members common for all DXF drawable entities. */
        int id_code;
                /*!< group code = 5\n
                 * Identification number for the entity.\n
                 * This is to be an unique (sequential) number in the DXF
                 * file. */
        char *linetype;
                /*!< group code = 6\n
                 * The linetype of the entity.\n
                 * Defaults to \c BYLAYER if ommitted in the DXF file. */
        char *layer;
                /*!< group code = 8\n
                 * Layer on which the entity is drawn.\n
                 * Defaults to layer "0" if no valid layername is given. */
        double thickness;
                /*!< group code = 39\n
                 * Thickness of the arc in the local Z-direction.\n
                 * Defaults to 0.0 if ommitted in the DXF file. */
        int color;
                /*!< group code = 62\n
                 * Color of the entity.\n
                 * Defaults to \c BYLAYER if ommitted in the DXF file.\n
                 * Note that entities encapsulated in a block with the
                 * color \c BYBLOCK are represented in the "native" color of
                 * the \c BLOCK entity. */
        int paperspace;
                /*!< group code = 67\n
                 * Entities are to be drawn on either \c PAPERSPACE or
                 * \c MODELSPACE.\n
                 * Optional, defaults to \c DXF_MODELSPACE (0). */
        /* Specific members for a DXF dimension. */
        char *dim_text;
                /*!< Dimension text explicitly entered by the user.\n
                 * If null or "<>", the dimension measurement is drawn
                 * as the text, if " " [one blank space], the text is
                 * suppressed.\n
                 * Anything else is drawn as the text.\n
                 * Group code = 1. */
        char *dimblock_name;
                /*!< Name of pseudo-Block containing the current
                 * dimension entity geometry.\n
                 * Group code = 2. */
        char *dimstyle_name;
                /*!< Dimension style name.\n
                 * Group code = 3.  */
        double x0;
                /*!< X-value of the definition point for all dimension
                 * types.\n
                 * Group code = 10. */
        double y0;
                /*!< Y-value of the definition point for all dimension
                 * types.\n
                 * Group code = 20. */
        double z0;
                /*!< Z-value of the definition point for all dimension
                 * types.\n
                 * Group code = 30. */
        double x1;
                /*!< X-value of the middle point of dimension text.\n
                 * Group code = 11. */
        double y1;
                /*!< Y-value of the middle point of dimension text.\n
                 * Group code = 21. */
        double z1;
                /*!< Z-value of the middle point of dimension text.\n
                 * Group code = 31. */
        double x2;
                /*!< X-value of the dimension block translation vector.\n
                 * Group code = 12. */
        double y2;
                /*!< Y-value of the dimension block translation vector.\n
                 * Group code = 22. */
        double z2;
                /*!< Z-value of the dimension block translation vector.\n
                 * Group code = 32. */
        double x3;
                /*!< X-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 13. */
        double y3;
                /*!< Y-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 23. */
        double z3;
                /*!< Z-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 33. */
        double x4;
                /*!< X-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 14. */
        double y4;
                /*!< Y-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 24. */
        double z4;
                /*!< Z-value of the definition point for linear and
                 * angular dimensions.\n
                 * Group code = 34. */
        double x5;
                /*!< X-value of the definition point for diameter,
                 * radius, and angular dimensions.\n
                 * Group code = 15. */
        double y5;
                /*!< Y-value of the definition point for diameter,
                 * radius, and angular dimensions.\n
                 * Group code = 25. */
        double z5;
                /*!< Z-value of the definition point for diameter,
                 * radius, and angular dimensions.\n
                 * Group code = 35. */
        double x6;
                /*!< X-value of the point defining dimension arc for
                 * angular dimensions.\n
                 * Group code = 16. */
        double y6;
                /*!< Y-value of the point defining dimension arc for
                 * angular dimensions.\n
                 * Group code = 26. */
        double z6;
                /*!< Z-value of the point defining dimension arc for
                 * angular dimensions.\n
                 * Group code = 36. */
        double leader_length;
                /*!< Leader length for radius and diameter dimensions.\n
                 * Group code = 40. */
        double text_line_spacing_factor;
                /*!< Dimension text line spacing factor (optional):\n
                 * Percentage of default (3-on-5) line spacing to be
                 * applied.\n
                 * Valid values range from 0.25 to 4.00.\n
                 * Group code = 41. */
        double actual_measurement;
                /*!< Actual measurement (optional; read-only value).\n
                 * Group code = 42. */
        double angle;
                /*!< Angle of rotated, horizontal, or vertical linear
                 * dimensions.\n
                 * Group code = 50. */
        double hor_dir;
                /*!< In addition, all dimension types have an optional
                 * group (group code 51) that indicates the horizontal
                 * direction for the Dimension entity.\n
                 * This determines the orientation of dimension text and
                 * dimension lines for horizontal, vertical, and rotated
                 * linear dimensions.\n
                 * The group value is the negative of the Entity
                 * Coordinate Systems (ECS) angle of the UCS X axis in
                 * effect when the Dimension was drawn.\n
                 * The X axis of the UCS in effect when the Dimension
                 * was drawn is always parallel to the XY plane for the
                 * Dimension's ECS, and the angle between the UCS X axis
                 * and the ECS X axis is a single 2D angle.\n
                 * The value in group code 51 is the angle from
                 * horizontal (the effective X axis) to the ECS X axis.\n
                 * Entity Coordinate Systems (ECS) are described
                 * elsewhere in this documentation.\n
                 * Group code = 51. */
        double obl_angle;
                /*!< Linear dimension types with an oblique angle have
                 * an optional group (group code 52).\n
                 * When added to the rotation angle of the linear
                 * dimension (group code 50) this gives the angle of the
                 * extension lines.\n
                 * Group code = 52. */
        double text_angle;
                /*!< The optional group code 53 is the rotation angle of
                 * the dimension text away from its default orientation
                 * (the direction of the dimension line).\n
                 * Group code = 53. */
        int flag;
                /*!< Dimension type.\n
                 * Values 0–6 are integer values that represent the
                 * dimension type.\n
                 * Values 32, 64, and 128 are bit values, which are
                 * added to the integer values (value 32 is always set
                 * in R13 and later releases).\n
                 * 0 = Rotated, horizontal, or vertical.\n
                 * 1 = Aligned.\n
                 * 2 = Angular.\n
                 * 3 = Diameter.\n
                 * 4 = Radius.\n
                 * 5 = Angular 3-point.\n
                 * 6 = Ordinate.\n
                 * 32 = Indicates that the block reference (group code
                 *      2) is referenced by this dimension only.\n
                 * 64 = Ordinate type.\n
                 *      This is a bit value (bit 7) used only with
                 *      integer value 6.\n
                 *      If set, ordinate is X-type; if not set, ordinate
                 *      is Y-type.\n
                 * 128 = This is a bit value (bit 8) added to the other
                 *       group 70 values if the dimension text has been
                 *       positioned at a user-defined location rather
                 *       than at the default location.\n
                 * Group code = 70. */
        int attachment_point;
                /*!< Attachment point:\n
                 * 1 = Top left;\n
                 * 2 = Top center;\n
                 * 3 = Top right;\n
                 * 4 = Middle left;\n
                 * 5 = Middle center;\n
                 * 6 = Middle right;\n
                 * 7 = Bottom left;\n
                 * 8 = Bottom center;\n
                 * 9 = Bottom right.\n
                 * Group code = 71. */
        int text_line_spacing;
                /*!< Dimension text line spacing style (optional):\n
                 * 1 (or missing) = At least (taller characters will
                 *   override);\n
                 * 2 = Exact (taller characters will not override).\n
                 * Group code = 72. */
        double extr_x0;
                /*!< X-value of the extrusion vector.\n
                 * Defaults to 0.0 if ommitted in the DXF file.\n
                 * Group code = 210. */
        double extr_y0;
                /*!< Y-value of the extrusion vector.\n
                 * Defaults to 0.0 if ommitted in the DXF file.\n
                 * Group code = 220. */
        double extr_z0;
                /*!< Z-value of the extrusion vector.\n
                 * Defaults to 1.0 if ommitted in the DXF file.\n
                 * Group code = 230. */
        struct DxfDimension *next;
                /*!< pointer to the next DxfDimension.\n
                 * \c NULL in the last DxfDimension. */
} DxfDimension, * DxfDimensionPtr;


DxfDimension *dxf_dimension_new ();


#endif /* _DIM_H */


/* EOF */
